[WGSL] Use UniqueRefVector to simplify working with AST

djg opened the pull request:
#### 8b4319ba38ad79d0a796d5e1c0dab1efd8a1616d
<pre>
[WGSL] Use UniqueRefVector to simplify working with AST
<a href="https://bugs.webkit.org/show_bug.cgi?id=246560">https://bugs.webkit.org/show_bug.cgi?id=246560</a>
rdar://problem/101198284

Reviewed by Dean Jackson.

Use UniqueRefVector for lists of AST nodes to eliminate the need to unwrap
UniqueRef&lt;&gt; with .get().

Tested by TestWebKitAPI/Tests/WGSL/ParserTests.cpp

* Source/WebGPU/WGSL/AST/Attribute.h:
* Source/WebGPU/WGSL/AST/Decl.h:
(WGSL::AST::Decl::~Decl):
* Source/WebGPU/WGSL/AST/Expression.h:
* Source/WebGPU/WGSL/AST/Expressions/ArrayAccess.h:
* Source/WebGPU/WGSL/AST/Expressions/CallableExpression.h:
* Source/WebGPU/WGSL/AST/FunctionDecl.h:
* Source/WebGPU/WGSL/AST/GlobalDirective.h:
* Source/WebGPU/WGSL/AST/ShaderModule.h:
* Source/WebGPU/WGSL/AST/Statement.h:
* Source/WebGPU/WGSL/AST/Statements/CompoundStatement.h:
* Source/WebGPU/WGSL/AST/StructureDecl.h:
* Source/WebGPU/WGSL/AST/VariableDecl.h:
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseShader):
(WGSL::Parser&lt;Lexer&gt;::parseAttributes):
(WGSL::Parser&lt;Lexer&gt;::parseStructDecl):
(WGSL::Parser&lt;Lexer&gt;::parseVariableDecl):
(WGSL::Parser&lt;Lexer&gt;::parseVariableDeclWithAttributes):
(WGSL::Parser&lt;Lexer&gt;::parseFunctionDecl):
(WGSL::Parser&lt;Lexer&gt;::parseStatement):
(WGSL::Parser&lt;Lexer&gt;::parseCompoundStatement):
(WGSL::Parser&lt;Lexer&gt;::parseArgumentExpressionList):
* Source/WebGPU/WGSL/ParserPrivate.h:
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(TestWGSLAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/255597@main">https://commits.webkit.org/255597@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6133916fe9396326411e9757fa289db2d77b8e7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92982 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23554 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102695 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/162941 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96984 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2194 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30512 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85367 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98809 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98648 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1500 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79456 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28412 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83413 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71526 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36911 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17037 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34721 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18229 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3876 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38593 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40830 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40518 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37417 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebGPU/WGSL/AST/ShaderModule.h:
```diff
@@ -39,41 +39,43 @@ namespace WGSL::AST {
-            if (is<StructDecl>(decl.get())) {
+        for (size_t i = decls.size(); i > 0; --i) {
+            UniqueRef<Decl> decl = decls.takeLast();
+            if (is<StructDecl>(decl)) {
```
Any reason you changed this?

djg replied:
Because the UniqueRefVector change means that decl is now Decl& not UniqueRef<Decl>& and Decl is move only, I’m taking UniqueRef out of the vector and downcasting them. This is going away in a follow up.

djg committed:
[WGSL] Use UniqueRefVector to simplify working with AST
https://bugs.webkit.org/show_bug.cgi?id=246560
rdar://problem/101198284

Reviewed by Dean Jackson.

Use UniqueRefVector for lists of AST nodes to eliminate the need to unwrap
UniqueRef<> with .get().

Tested by TestWebKitAPI/Tests/WGSL/ParserTests.cpp

* Source/WebGPU/WGSL/AST/Attribute.h:
* Source/WebGPU/WGSL/AST/Decl.h:
(WGSL::AST::Decl::~Decl):
* Source/WebGPU/WGSL/AST/Expression.h:
* Source/WebGPU/WGSL/AST/Expressions/ArrayAccess.h:
* Source/WebGPU/WGSL/AST/Expressions/CallableExpression.h:
* Source/WebGPU/WGSL/AST/FunctionDecl.h:
* Source/WebGPU/WGSL/AST/GlobalDirective.h:
* Source/WebGPU/WGSL/AST/ShaderModule.h:
* Source/WebGPU/WGSL/AST/Statement.h:
* Source/WebGPU/WGSL/AST/Statements/CompoundStatement.h:
* Source/WebGPU/WGSL/AST/StructureDecl.h:
* Source/WebGPU/WGSL/AST/VariableDecl.h:
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseShader):
(WGSL::Parser<Lexer>::parseAttributes):
(WGSL::Parser<Lexer>::parseStructDecl):
(WGSL::Parser<Lexer>::parseVariableDecl):
(WGSL::Parser<Lexer>::parseVariableDeclWithAttributes):
(WGSL::Parser<Lexer>::parseFunctionDecl):
(WGSL::Parser<Lexer>::parseStatement):
(WGSL::Parser<Lexer>::parseCompoundStatement):
(WGSL::Parser<Lexer>::parseArgumentExpressionList):
* Source/WebGPU/WGSL/ParserPrivate.h:
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(TestWGSLAPI::TEST):

Canonical link: https://commits.webkit.org/255597@main
